### ISSUE TITLE
dagster-snowflake-pandas no longer quotes identifiers

### ISSUE THREAD

**Issue by TimoVink** [dagster-io/dagster#20552]
### Dagster version

1.6.11

### What's the issue?

We recently upgraded from v1.5.9 to v1.6.11. The Snowflake Pandas IO manager now fails to create tables and upload data in cases where the returned dataframe has a column name that is not a valid identifier unless it's quoted (such as `5_stars`, which is invalid because it starts with a number).

The change of behaviour was (accidentally?) introduced as part of: https://github.com/dagster-io/dagster/commit/45058bea4f8fe49d548eb1853ac5989ed4d1b0ae. Note how `quote_identifiers` is explicitly to `False` in `handle_output`.

This is an (undocumented) breaking change in behaviour from older versions.

### What did you expect to happen?

I expected the IO manager to continue quoting identifiers, as it did in previous versions.

### How to reproduce?

Configure an asset to use the Snowflake IO manager, and return a dataframe such as:
```python
df = pd.DataFrame({ 'name': ['Sample'], '5_stars': [True] })
```

Note that `5_stars` is not a valid column name in Snowflake (unless quoted).

### Deployment type

Dagster Helm chart

### Deployment details

N/A

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

**Comment by TimoVink** [comment]
I will admit that having to quote identifiers is not great, and that _not_ quoting identifiers (the new behaviour) is a sensible default. 

However, even if a decision is made to stick with _not_ quoting identifiers by default, it'd be great if we could document this breaking change. It'd also be nice in this case if there was a way to opt-in to quoting for users who cannot easily rename their existing columns. Otherwise we might be stuck on v1.5 for a while.

Thanks for taking a look! 👍🏻

**Comment by sryza** [comment]
@jamiedemaria - are you familiar with this change?

**Comment by jamiedemaria** [comment]
yeah this is almost 100% due to switching to `write_pandas`. Previously we were use `to_sql` with snowflake's `pd_writer` which did not use quote identifiers for table and schema names, but i guess did quote column names (I was unaware of this and still need to find evidence to support that this is the case). So when we switched to `write_pandas` I set `quote_identifiers` to False to match the behavior that `to_sql` has for table names.

**Comment by TimoVink** [comment]
Thanks for taking a look!

Let me try an dig up some extra information I'll try creating and inserting some dummy data both before and after that change and get the exact SQL sent to Snowflake. Back in a bit!

**Comment by jamiedemaria** [comment]
that'd be awesome, thank you!

**Comment by TimoVink** [comment]
Sorry should've included this in the original report. Here's my setup, comparing `1.5.14` with `1.6.11`

Environment is like so (and respectively for 1.6.11):
```
python = "3.10.6"
dagster = "1.5.14"
dagster-webserver = "1.5.14"
dagster-snowflake = "0.21.14"
dagster-snowflake-pandas = "0.21.14"
```

Test code as follows (insert your own SF config of course)
```python
from dagster import asset, Definitions
import pandas as pd
from dagster_snowflake_pandas import SnowflakePandasIOManager

@asset my_asset():
    return pd.DataFrame({ "name": ["Test"], "5_stars": [True] })

Definitions(assets=[my_asset], resources={ "io_manager": SnowflakePandasIOManager(...) }
```

On **1.5.14** I see this SQL on materialization:
```sql
DESC TABLE my_asset;

-- I assume this only runs if the above indicates the table does not exist
CREATE TABLE my_asset (
    "NAME" TEXT,
    "5_STARS" BOOLEAN
);

CREATE TEMP STAGE "wygcpwnngcv" FILE_FORMAT=(TYPE=PARQUET COMPRESSION=GZIP);

PUT file:///tmp/tmprte6wc6h/file0.txt @"wygcpwnngcv" PARALLEL=4;

COPY INTO "MY_ASSET" ("NAME", "5_STARS") 
FROM (SELECT $1:"NAME",$1:"5_STARS" FROM @"wygcpwnngcv")
FILE_FORMAT=(TYPE=PARQUET COMPRESSION=auto) PURGE=TRUE ON_ERROR=abort_statement;
```

On **1.6.11** I see this SQL on materialization (regardless of whether or not the table already exists):
```sql
CREATE TABLE IF NOT EXISTS analytics.public.MY_ASSET (NAME TEXT, 5_STARS BOOLEAN);
-- ERROR: SQL compilation error: Syntax error line 1 at position 66 unexpected '5'.
```

**Comment by TimoVink** [comment]
So as far as I can tell the differences are:
1. The table name is now uppercased, but since it's unquoted this shouldn't make a difference
2. The column names are no longer quoted -- this is the part that breaks things for me.

Just by visual inspect I would think that setting `quote_identifiers=True` might gives us the same behaviour we had previously?
https://github.com/dagster-io/dagster/blob/6622cbd5f07c75165f82b2f41d4e81f6f1972064/python_modules/libraries/dagster-snowflake-pandas/dagster_snowflake_pandas/snowflake_pandas_type_handler.py#L128

**Comment by jamiedemaria** [comment]
we might have to force upper casing on the database and schema as well. I need to refresh my knowledge of how snowflake handlers non-quoted identifiers. If it's as simple as uppercasing those, then that would be a good solution to maintain consistency until we add support for quoting database/schema/table

**Comment by TimoVink** [comment]
I tried monkeypatching `quote_identifiers=False` and indeed immediately ran into that: It then quotes the database name as well. Which in my example is _not_ what I want ...very annoying.

I did find a ~workaround~ complete hack to get the behaviour I'm looking for. I replaced 

https://github.com/dagster-io/dagster/blob/6622cbd5f07c75165f82b2f41d4e81f6f1972064/python_modules/libraries/dagster-snowflake-pandas/dagster_snowflake_pandas/snowflake_pandas_type_handler.py#L107

With:
```python
with_uppercase_cols = obj.rename(lambda s: f"\"{s.upper()}\"", copy=False, axis="columns")
```
To essentially manually quote the column names, while leaving the database and schema name untouched.

**Comment by TimoVink** [comment]
If someone else is reading along and is okay using a complete hack in the interim, here's what does the trick for me:

```python
class MyTypeHandler(SnowflakePandasTypeHandler):
    def handle_output(self, context: OutputContext, table_slice: TableSlice, obj: pd.DataFrame, connection) -> Mapping[str, RawMetadataValue]:
        quoted_df = obj.rename(lambda s: f"\"{s}\"", copy=False, axis="columns")
        return super().handle_output(context, table_slice, quoted_df, connection)


class MyIOManager(SnowflakePandasIOManager):
    @staticmethod
    def type_handlers() -> Sequence[DbTypeHandler]:
        return [MyTypeHandler()]
```
And then use `MyIOManager` as a drop-in for `SnowflakePandasIOManager`.

**Comment by jamiedemaria** [comment]
thanks for putting this all together! this is super helpful for me too. i'm going to see about manually capitalizing the database/schema/etc and look at the feasibility of merging that

**Comment by jamiedemaria** [comment]
@TimoVink  I put together this PR that I think should fix the issue https://github.com/dagster-io/dagster/pull/20631

**Comment by TimoVink** [comment]
Thanks so much for taking a look!

In meetings all morning, but I'll give the branch a try on my dummy tests and real data this afternoon if I find some time.

**Comment by fcc-loganallen** [comment]
This change hasn't been merged yet, correct? Ran into the same issue.
